python/lsst/meas/extensions/shapeHSM/_hsm_moments.py

```diff
@@ -0,0 +1,529 @@
+import lsst.pex.config as pexConfig
+import numpy as np
+from lsst.geom import Point2D, Point2I
+from lsst.pex.exceptions import InvalidParameterError, NotFoundError
+
+
+class HsmMomentsConfig(measBase.SingleFramePluginConfig):
+    """Base configuration for HSM adaptive moments measurement."""
+
+    roundMoments = pexConfig.Field[bool](doc="Use round weight function?", default=False)
+    addFlux = pexConfig.Field[bool](doc="Store measured flux?", default=False)
+    subtractCenter = pexConfig.Field[bool](doc="Subtract starting center from x/y outputs?", default=False)
+
+
+class HsmMomentsPlugin(measBase.SingleFramePlugin):
+    """Base plugin for HSM adaptive moments measurement."""
+
+    ConfigClass = HsmMomentsConfig
+
+    def __init__(self, config, name, schema, metadata, logName=None):
+        if logName is None:
+            logName = __name__
+        super().__init__(config, name, schema, metadata, logName=logName)
+
+        # Define flags for possible issues that might arise during measurement.
+        flagDefs = measBase.FlagDefinitionList()
+        self.FAILURE = flagDefs.addFailureFlag("General failure flag, set if anything went wrong")
+        self.NO_PIXELS = flagDefs.add("flag_no_pixels", "No pixels to measure")
+        self.NOT_CONTAINED = flagDefs.add(
+            "flag_not_contained", "Center not contained in footprint bounding box"
+        )
+        self.PARENT_SOURCE = flagDefs.add("flag_parent_source", "Parent source, ignored")
+        self.GALSIM = flagDefs.add("flag_galsim", "GalSim failure")
+        self.INVALID_PARAM = flagDefs.add("flag_invalid_param", "Invalid combination of moments")
+        self.EDGE = flagDefs.add("flag_edge", "Variance undefined outside image edge")
+        self.NO_PSF = flagDefs.add("flag_no_psf", "Exposure lacks PSF")
+
+        # Embed the flag definitions in the schema using a flag handler.
+        self.flagHandler = measBase.FlagHandler.addFields(schema, name, flagDefs)
+
+        # Utilize a safe centroid extractor that uses the detection footprint
+        # as a fallback if necessary.
+        self.centroidExtractor = measBase.SafeCentroidExtractor(schema, name)
+        self.log = logging.getLogger(self.logName)
+
+    @classmethod
+    def getExecutionOrder(cls):
+        return cls.SHAPE_ORDER
+
+    def _calculate(
+        self,
+        record: afwTable.SourceRecord,
+        *,
+        image: galsim.Image,
+        weight: galsim.Image | None,
+        badpix: galsim.Image | None,
+        sigma: float,
+        precision: float = 1.0e-6,
+        centroid: Point2D,
+    ):
+        """
+        Calculate adaptive moments using GalSim's HSM and modify the record in
+        place.
+
+        Parameters
+        ----------
+        record : `~lsst.afw.table.SourceRecord`
+            Record to store measurements.
+        image : `~galsim.Image`
+            Image on which to perform measurements.
+        weight : `~galsim.Image`
```

Comment:
```suggestion
        weight : `~galsim.Image`, optional
```
and set the default to `None` if it's allowed. Similarly for the other ones that accept `None`.  That'd mean you need to move up the `sigma` argument.

Reply:
Done! I used all the defaults from `galsim` including for `sigma`, so just kept the same order.